File: codes/kotlin/p970.kt
import kotlin.math.*
import kotlin.complex.Complex

fun lambertW(z: Complex, k: Int): Complex {
    var w = Complex(ln(z.abs()), z.phase + 2*PI*k)
    repeat(50) {
        val ew = w.exp()
        w -= (w*ew - z)/(ew*(w + Complex(1.0,0.0)))
    }
    return w
}

File: codes/Haskell/p970.hs
import Data.Complex

lambertW :: Complex Double -> Int -> Complex Double
lambertW z k = iterateNewton w0 !! 30
  where
    r = magnitude z
    t = phase z
    w0 = log r :+ (t + 2*pi*fromIntegral k)
    iterateNewton w =
      let ew = exp w
          f  = w*ew - z
          fp = ew*(w+1)
      in w - f/fp
